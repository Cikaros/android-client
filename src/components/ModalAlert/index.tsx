import React, { useEffect } from 'react';
import { Image, Modal, Text, TouchableWithoutFeedback, View, ViewStyle } from 'react-native';
import Button from '../Button/index';

interface Props {
  title?: string;
  text?: string;
  visible: boolean;
  autoClose?: boolean;
  textOkButton?: string;
  onOk?: () => void;
  textCancelButton?: string;
  onCancel?: () => void;
  iconSource?: any;
  body?: () => JSX.Element;
  messageContent?: () => JSX.Element;
  onPressOverlay?: () => void | undefined;
  onPress?: () => void;
}

const styles = {
  NbModalCard: {
    padding: 24,
    backgroundColor: '#FFF',
    marginLeft: 22,
    marginRight: 22,
    borderRadius: 14,
  } as ViewStyle,
  NbModalBackdrop: {
    flex: 1,
    backgroundColor: 'rgba(0, 0, 0, 0.3)',
    justifyContent: 'center',
  } as ViewStyle,
  NbModalContent: {
    flexDirection: 'column',
    alignItems: 'center',
    marginBottom: 24,
    paddingTop: 17,
  } as ViewStyle,
  NbModalBody: {
    paddingHorizontal: 0,
    paddingVertical: 15,
  } as ViewStyle,
  NbModalTextTitle: {
    fontSize: 18,
    fontWeight: '700',
    lineHeight: 24,
    marginBottom: 8,
    color: 'rgba(0, 0, 0, 0.85)',
    textAlign: 'center',
  } as ViewStyle,
  NbModalBottomOptions: {
    flexDirection: 'column',
    gap: 15,
  } as ViewStyle,
  NbModalTextContent: {
    fontSize: 16,
    lineHeight: 22,
    color: 'rgba(77, 77, 77, 1)',
    textAlign: 'center',
  } as ViewStyle,
};

const ModalAlert: React.FC<Props> = ({
  title,
  text,
  visible,
  autoClose,
  textOkButton,
  onOk,
  textCancelButton,
  onCancel,
  iconSource,
  body,
  messageContent,
  onPressOverlay,
}) => {
  useEffect(() => {
    if (visible && autoClose) {
      setTimeout(() => {
        onPressOverlay && onPressOverlay();
      }, 5000);
    }
  }, [visible]);

  return (
    <Modal visible={visible} transparent={true} animationType='fade'>
      <TouchableWithoutFeedback id='overlayId' onPress={onPressOverlay}>
        <View style={styles.NbModalBackdrop}>
          <View style={styles.NbModalCard}>
            <TouchableWithoutFeedback>
              <View style={styles.NbModalContent}>
                {iconSource && (
                  <Image
                    source={iconSource}
                    style={{
                      marginRight: 5,
                      marginBottom: 15,
                      resizeMode: 'contain',
                    }}
                  />
                )}
                {!body ? (
                  <View style={styles.NbModalBody}>
                    <Text style={styles.NbModalTextTitle}>{title}</Text>
                    {text && <Text style={styles.NbModalTextContent}>{text}</Text>}
                    {messageContent && !text && messageContent()}
                  </View>
                ) : (
                  body()
                )}
              </View>
            </TouchableWithoutFeedback>
            <View style={styles.NbModalBottomOptions}>
              <Button
                onPress={() => onOk && onOk()}
                text={textOkButton || 'Ok'}
                id='okButton'
              />
              {onCancel && (
                <Button
                  onPress={() => onCancel()}
                  text={textCancelButton || 'Cancel'}
                  outline={true}
                  id='cancelButton'
                />
              )}
            </View>
          </View>
        </View>
      </TouchableWithoutFeedback>
    </Modal>
  );
};

export default React.memo(ModalAlert);
