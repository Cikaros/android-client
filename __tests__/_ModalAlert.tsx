import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';
import ModalAlert from '../src/components/ModalAlert/index';

interface Props {
  title?: string;
  text?: string;
  visible: boolean;
  autoClose: boolean;
  textOkButton?: string;
  onOk?: () => void;
  textCancelButton?: string;
  onCancel?: () => void;
  icon?: any;
  iconSource?: any;
  body?: () => JSX.Element;
  messageContent?: () => JSX.Element;
  onPressOverlay?: () => void;
  onPress?: () => void;
}

describe('ModalAlert', () => {
  const mockProps: Props = {
    title: 'Test Title',
    text: 'Test Text',
    visible: false,
    autoClose: false,
    textOkButton: 'OK',
    onOk: jest.fn(),
    textCancelButton: 'Cancel',
    onCancel: jest.fn(),
    icon: 'test-icon',
    iconSource: 'test-icon-source',
    body: jest.fn(),
    messageContent: jest.fn(),
    onPressOverlay: jest.fn(),
    onPress: jest.fn(),
  };

  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<ModalAlert {...mockProps} />);
  });
  // Suppress the warning message
  console.error = jest.fn();

  it('renders correctly', () => {
    const wrapper = shallow(<ModalAlert {...mockProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls onOk when OK button is pressed', () => {
    const wrapper = shallow(<ModalAlert {...mockProps} />);
    wrapper.find('[id="okButton"]').simulate('press');
    expect(mockProps.onOk).toHaveBeenCalled();
  });

  it('calls onCancel when Cancel button is pressed', () => {
    const onCancelMock = jest.fn();
    const wrapper = shallow(
      <ModalAlert {...mockProps} onCancel={onCancelMock} />,
    );
    wrapper.find('[id="cancelButton"]').simulate('press');
    expect(onCancelMock).toHaveBeenCalled();
  });

  it('does not throw an error when Cancel button is not provided', () => {
    const wrapper = shallow(<ModalAlert {...mockProps} />);
    expect(() => {
      wrapper.find('[id="cancelButton"]').simulate('press');
    }).not.toThrow();
  });

  it('calls onPressOverlay when overlay is pressed', () => {
    const wrapper = shallow(<ModalAlert {...mockProps} />);
    wrapper.find('[id="overlayId"]').first().simulate('press');
    expect(mockProps.onPressOverlay).toMatchSnapshot();
  });
});
