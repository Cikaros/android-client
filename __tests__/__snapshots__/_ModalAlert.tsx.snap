// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalAlert calls onPressOverlay when overlay is pressed 1`] = `
[MockFunction] {
  "calls": [
    [],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`ModalAlert renders correctly 1`] = `ShallowWrapper {}`;
